// @google-cloud/vertexai/build/src/genai/client.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Client = exports.SDK_VERSION = void 0;
/**
 * @license
 * Copyright 2026 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */
const vertex_internal_1 = require("@google/genai/vertex_internal");
const agentengines_1 = require("./agentengines");
exports.SDK_VERSION = '1.12.0'; // x-release-please-version
let agentEnginesInternalWarned = false;
class Client {
    constructor(options) {
        const auth = new vertex_internal_1.NodeAuth({
            googleAuthOptions: {
                scopes: ['https://www.googleapis.com/auth/cloud-platform'],
            },
        });
        const uploader = new vertex_internal_1.NodeUploader();
        const downloader = new vertex_internal_1.NodeDownloader();
        this.apiClient = new vertex_internal_1.ApiClient({
            auth,
            uploader,
            downloader,
            project: options.project,
            location: options.location,
            vertexai: true,
            httpOptions: options.apiEndpoint ? { baseUrl: options.apiEndpoint } :
                undefined,
            userAgentExtra: `vertex-genai-modules/${exports.SDK_VERSION}`,
        });
        this._agentEnginesInternal = new agentengines_1.AgentEngines(this.apiClient);
    }
    /**
     * Getter for agentEnginesInternal that logs a warning on first use.
     */
    get agentEnginesInternal() {
        if (!agentEnginesInternalWarned) {
            console.warn('The agentEnginesInternal implementation is experimental, and may change in future versions.');
            agentEnginesInternalWarned = true;
        }
        return this._agentEnginesInternal;
    }
}
exports.Client = Client;
