// @genkit-ai/core/lib/config.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var config_exports = {};
__export(config_exports, {
  getGenkitRuntimeConfig: () => getGenkitRuntimeConfig,
  resetGenkitRuntimeConfig: () => resetGenkitRuntimeConfig,
  setGenkitRuntimeConfig: () => setGenkitRuntimeConfig
});
module.exports = __toCommonJS(config_exports);
var import_logging = require("./logging");
const CONFIG_KEY = "__GENKIT_RUNTIME_CONFIG__";
function getConfig() {
  if (!global[CONFIG_KEY]) {
    global[CONFIG_KEY] = {};
  }
  return global[CONFIG_KEY];
}
function setGenkitRuntimeConfig(config) {
  if (config.jsonSchemaMode === "interpret") {
    import_logging.logger.warn(
      "It looks like you're trying to disable schema code generation. Please ensure that the '@cfworker/json-schema' package is installed: `npm i --save @cfworker/json-schema`"
    );
  }
  const current = getConfig();
  global[CONFIG_KEY] = { ...current, ...config };
}
function getGenkitRuntimeConfig() {
  const config = getConfig();
  return {
    jsonSchemaMode: config.jsonSchemaMode ?? "compile",
    sandboxedRuntime: config.sandboxedRuntime ?? false
  };
}
function resetGenkitRuntimeConfig() {
  global[CONFIG_KEY] = {};
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  getGenkitRuntimeConfig,
  resetGenkitRuntimeConfig,
  setGenkitRuntimeConfig
});
