// @genkit-ai/core/lib/dynamic-action-provider.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var dynamic_action_provider_exports = {};
__export(dynamic_action_provider_exports, {
  defineDynamicActionProvider: () => defineDynamicActionProvider,
  isDynamicActionProvider: () => isDynamicActionProvider
});
module.exports = __toCommonJS(dynamic_action_provider_exports);
var z = __toESM(require("zod"));
var import_action = require("./action.js");
var import_error = require("./error.js");
class SimpleCache {
  value;
  expiresAt;
  ttlMillis;
  dap;
  dapFn;
  fetchPromise = null;
  constructor(config, dapFn) {
    this.dapFn = dapFn;
    this.ttlMillis = !config.cacheConfig?.ttlMillis ? 3 * 1e3 : config.cacheConfig?.ttlMillis;
  }
  setDap(dap) {
    this.dap = dap;
  }
  setValue(value) {
    const dapId = this.dap?.__action?.name;
    if (dapId) {
      Object.values(value).forEach((actionList) => {
        actionList?.forEach((action) => {
          action.__action.key = `/dynamic-action-provider/${dapId}:${action.__action.actionType}/${action.__action.name}`;
        });
      });
    }
    this.value = value;
    this.expiresAt = Date.now() + this.ttlMillis;
  }
  /**
   * Gets or fetches the DAP data.
   * @param skipTrace Don't run the action. i.e. don't create a trace log.
   * @returns The DAP data
   */
  async getOrFetch(params) {
    const isStale = !this.value || !this.expiresAt || this.ttlMillis < 0 || Date.now() > this.expiresAt;
    if (!isStale) {
      return this.value;
    }
    if (!this.fetchPromise) {
      this.fetchPromise = (async () => {
        try {
          if (this.dap && !params?.skipTrace) {
            await this.dap.run();
          } else {
            this.setValue(await this.dapFn());
          }
          if (!this.value) {
            throw new Error("value is undefined");
          }
          return this.value;
        } catch (error) {
          console.error("Error fetching Dynamic Action Provider value:", error);
          this.invalidate();
          throw error;
        } finally {
          this.fetchPromise = null;
        }
      })();
    }
    return await this.fetchPromise;
  }
  invalidate() {
    this.value = void 0;
  }
}
function isDynamicActionProvider(obj) {
  return obj.__action?.metadata?.type == "dynamic-action-provider";
}
function transformDapValue(value) {
  return Object.values(value).flatMap(
    (actions) => actions?.map((a) => a.__action) || []
  );
}
function defineDynamicActionProvider(registry, config, fn) {
  let cfg;
  if (typeof config == "string") {
    cfg = { name: config };
  } else {
    cfg = { ...config };
  }
  const cache = new SimpleCache(cfg, fn);
  const a = (0, import_action.defineAction)(
    registry,
    {
      ...cfg,
      inputSchema: z.void(),
      outputSchema: z.array(import_action.ActionMetadataSchema),
      actionType: "dynamic-action-provider",
      metadata: { ...cfg.metadata || {}, type: "dynamic-action-provider" }
    },
    async (_options) => {
      const dapValue = await fn();
      cache.setValue(dapValue);
      return transformDapValue(dapValue);
    }
  );
  implementDap(a, cache);
  return a;
}
function implementDap(dap, cache) {
  cache.setDap(dap);
  dap.__cache = cache;
  dap.invalidateCache = () => {
    dap.__cache.invalidate();
  };
  dap.getAction = async (actionType, actionName) => {
    const result = await dap.__cache.getOrFetch();
    if (result[actionType]) {
      return result[actionType].find((t) => t.__action.name == actionName);
    }
    return void 0;
  };
  dap.listActionMetadata = async (actionType, actionName) => {
    const result = await dap.__cache.getOrFetch();
    if (!result[actionType]) {
      return [];
    }
    const metadata = result[actionType].map((a) => a.__action);
    if (actionName == "*") {
      return metadata;
    }
    if (actionName.endsWith("*")) {
      const prefix = actionName.slice(0, -1);
      return metadata.filter((m) => m.name.startsWith(prefix));
    }
    return metadata.filter((m) => m.name == actionName);
  };
  dap.getActionMetadataRecord = async (dapPrefix) => {
    const dapActions = {};
    const result = await dap.__cache.getOrFetch({ skipTrace: true });
    for (const actions of Object.values(result)) {
      const metadataList = actions.map((a) => a.__action);
      for (const metadata of metadataList) {
        if (!metadata.name || !metadata.key) {
          throw new import_error.GenkitError({
            status: "INVALID_ARGUMENT",
            message: `Invalid metadata when listing dynamic actions from ${dapPrefix} - name required`
          });
        }
        dapActions[metadata.key] = metadata;
      }
    }
    return dapActions;
  };
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  defineDynamicActionProvider,
  isDynamicActionProvider
});
