// @genkit-ai/core/lib/reflection-types.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var reflection_types_exports = {};
__export(reflection_types_exports, {
  ReflectionCancelActionParamsSchema: () => ReflectionCancelActionParamsSchema,
  ReflectionCancelActionResponseSchema: () => ReflectionCancelActionResponseSchema,
  ReflectionConfigureParamsSchema: () => ReflectionConfigureParamsSchema,
  ReflectionEndInputStreamParamsSchema: () => ReflectionEndInputStreamParamsSchema,
  ReflectionListActionsResponseSchema: () => ReflectionListActionsResponseSchema,
  ReflectionListValuesParamsSchema: () => ReflectionListValuesParamsSchema,
  ReflectionListValuesResponseSchema: () => ReflectionListValuesResponseSchema,
  ReflectionRegisterParamsSchema: () => ReflectionRegisterParamsSchema,
  ReflectionRunActionParamsSchema: () => ReflectionRunActionParamsSchema,
  ReflectionRunActionStateParamsSchema: () => ReflectionRunActionStateParamsSchema,
  ReflectionSendInputStreamChunkParamsSchema: () => ReflectionSendInputStreamChunkParamsSchema,
  ReflectionStreamChunkParamsSchema: () => ReflectionStreamChunkParamsSchema
});
module.exports = __toCommonJS(reflection_types_exports);
var import_zod = require("zod");
var import_action = require("./action.js");
const ReflectionRegisterParamsSchema = import_zod.z.object({
  id: import_zod.z.string(),
  pid: import_zod.z.number(),
  name: import_zod.z.string().optional(),
  genkitVersion: import_zod.z.string().optional(),
  reflectionApiSpecVersion: import_zod.z.number().optional(),
  envs: import_zod.z.array(import_zod.z.string()).optional()
});
const ReflectionStreamChunkParamsSchema = import_zod.z.object({
  requestId: import_zod.z.string(),
  chunk: import_zod.z.any()
});
const ReflectionRunActionStateParamsSchema = import_zod.z.object({
  requestId: import_zod.z.string(),
  state: import_zod.z.object({
    traceId: import_zod.z.string().optional()
  }).optional()
});
const ReflectionConfigureParamsSchema = import_zod.z.object({
  telemetryServerUrl: import_zod.z.string().optional()
});
const ReflectionListValuesParamsSchema = import_zod.z.object({
  type: import_zod.z.string()
});
const ReflectionRunActionParamsSchema = import_zod.z.object({
  key: import_zod.z.string(),
  input: import_zod.z.any().optional(),
  context: import_zod.z.any().optional(),
  telemetryLabels: import_zod.z.record(import_zod.z.string(), import_zod.z.string()).optional(),
  stream: import_zod.z.boolean().optional(),
  streamInput: import_zod.z.boolean().optional()
});
const ReflectionCancelActionParamsSchema = import_zod.z.object({
  traceId: import_zod.z.string()
});
const ReflectionSendInputStreamChunkParamsSchema = import_zod.z.object({
  requestId: import_zod.z.string(),
  chunk: import_zod.z.any()
});
const ReflectionEndInputStreamParamsSchema = import_zod.z.object({
  requestId: import_zod.z.string()
});
const ReflectionListActionsResponseSchema = import_zod.z.object({
  actions: import_zod.z.record(import_zod.z.string(), import_action.ActionMetadataSchema)
});
const ReflectionListValuesResponseSchema = import_zod.z.object({
  values: import_zod.z.record(import_zod.z.any())
});
const ReflectionCancelActionResponseSchema = import_zod.z.object({
  message: import_zod.z.string()
});
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  ReflectionCancelActionParamsSchema,
  ReflectionCancelActionResponseSchema,
  ReflectionConfigureParamsSchema,
  ReflectionEndInputStreamParamsSchema,
  ReflectionListActionsResponseSchema,
  ReflectionListValuesParamsSchema,
  ReflectionListValuesResponseSchema,
  ReflectionRegisterParamsSchema,
  ReflectionRunActionParamsSchema,
  ReflectionRunActionStateParamsSchema,
  ReflectionSendInputStreamChunkParamsSchema,
  ReflectionStreamChunkParamsSchema
});
