// @genkit-ai/core/lib/reflection-types.mjs
import { z } from "zod";
import { ActionMetadataSchema } from "./action.js";
const ReflectionRegisterParamsSchema = z.object({
  id: z.string(),
  pid: z.number(),
  name: z.string().optional(),
  genkitVersion: z.string().optional(),
  reflectionApiSpecVersion: z.number().optional(),
  envs: z.array(z.string()).optional()
});
const ReflectionStreamChunkParamsSchema = z.object({
  requestId: z.string(),
  chunk: z.any()
});
const ReflectionRunActionStateParamsSchema = z.object({
  requestId: z.string(),
  state: z.object({
    traceId: z.string().optional()
  }).optional()
});
const ReflectionConfigureParamsSchema = z.object({
  telemetryServerUrl: z.string().optional()
});
const ReflectionListValuesParamsSchema = z.object({
  type: z.string()
});
const ReflectionRunActionParamsSchema = z.object({
  key: z.string(),
  input: z.any().optional(),
  context: z.any().optional(),
  telemetryLabels: z.record(z.string(), z.string()).optional(),
  stream: z.boolean().optional(),
  streamInput: z.boolean().optional()
});
const ReflectionCancelActionParamsSchema = z.object({
  traceId: z.string()
});
const ReflectionSendInputStreamChunkParamsSchema = z.object({
  requestId: z.string(),
  chunk: z.any()
});
const ReflectionEndInputStreamParamsSchema = z.object({
  requestId: z.string()
});
const ReflectionListActionsResponseSchema = z.object({
  actions: z.record(z.string(), ActionMetadataSchema)
});
const ReflectionListValuesResponseSchema = z.object({
  values: z.record(z.any())
});
const ReflectionCancelActionResponseSchema = z.object({
  message: z.string()
});
export {
  ReflectionCancelActionParamsSchema,
  ReflectionCancelActionResponseSchema,
  ReflectionConfigureParamsSchema,
  ReflectionEndInputStreamParamsSchema,
  ReflectionListActionsResponseSchema,
  ReflectionListValuesParamsSchema,
  ReflectionListValuesResponseSchema,
  ReflectionRegisterParamsSchema,
  ReflectionRunActionParamsSchema,
  ReflectionRunActionStateParamsSchema,
  ReflectionSendInputStreamChunkParamsSchema,
  ReflectionStreamChunkParamsSchema
};
