// @google-cloud/vertexai/build/src/genai/converters/_agentengines_converters.js
"use strict";
/**
 * @license
 * Copyright 2025 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.updateAgentEngineRequestParametersToVertex = exports.updateAgentEngineConfigToVertex = exports.structuredMemorySchemaConfigToVertex = exports.structuredMemorySchemaConfigFromVertex = exports.structuredMemoryConfigToVertex = exports.structuredMemoryConfigFromVertex = exports.reasoningEngineFromVertex = exports.reasoningEngineContextSpecToVertex = exports.reasoningEngineContextSpecMemoryBankConfigToVertex = exports.reasoningEngineContextSpecMemoryBankConfigFromVertex = exports.reasoningEngineContextSpecFromVertex = exports.queryAgentEngineRequestParametersToVertex = exports.queryAgentEngineConfigToVertex = exports.listReasoningEnginesResponseFromVertex = exports.listAgentEngineRequestParametersToVertex = exports.listAgentEngineConfigToVertex = exports.getAgentEngineRequestParametersToVertex = exports.getAgentEngineOperationParametersToVertex = exports.deleteAgentEngineRequestParametersToVertex = exports.createAgentEngineRequestParametersToVertex = exports.createAgentEngineConfigToVertex = exports.agentEngineOperationFromVertex = void 0;
// Code generated by the Google Gen AI SDK generator DO NOT EDIT.
const common = require("@google/genai/vertex_internal");
function agentEngineOperationFromVertex(fromObject) {
    const toObject = {};
    const fromName = common.getValueByPath(fromObject, ['name']);
    if (fromName != null) {
        common.setValueByPath(toObject, ['name'], fromName);
    }
    const fromMetadata = common.getValueByPath(fromObject, ['metadata']);
    if (fromMetadata != null) {
        common.setValueByPath(toObject, ['metadata'], fromMetadata);
    }
    const fromDone = common.getValueByPath(fromObject, ['done']);
    if (fromDone != null) {
        common.setValueByPath(toObject, ['done'], fromDone);
    }
    const fromError = common.getValueByPath(fromObject, ['error']);
    if (fromError != null) {
        common.setValueByPath(toObject, ['error'], fromError);
    }
    const fromResponse = common.getValueByPath(fromObject, ['response']);
    if (fromResponse != null) {
        common.setValueByPath(toObject, ['response'], reasoningEngineFromVertex(fromResponse));
    }
    return toObject;
}
exports.agentEngineOperationFromVertex = agentEngineOperationFromVertex;
function createAgentEngineConfigToVertex(fromObject, parentObject) {
    const toObject = {};
    const fromDisplayName = common.getValueByPath(fromObject, ['displayName']);
    if (parentObject !== undefined && fromDisplayName != null) {
        common.setValueByPath(parentObject, ['displayName'], fromDisplayName);
    }
    const fromDescription = common.getValueByPath(fromObject, ['description']);
    if (parentObject !== undefined && fromDescription != null) {
        common.setValueByPath(parentObject, ['description'], fromDescription);
    }
    const fromSpec = common.getValueByPath(fromObject, ['spec']);
    if (parentObject !== undefined && fromSpec != null) {
        common.setValueByPath(parentObject, ['spec'], fromSpec);
    }
    const fromContextSpec = common.getValueByPath(fromObject, ['contextSpec']);
    if (parentObject !== undefined && fromContextSpec != null) {
        common.setValueByPath(parentObject, ['contextSpec'], reasoningEngineContextSpecToVertex(fromContextSpec));
    }
    const fromPscInterfaceConfig = common.getValueByPath(fromObject, [
        'pscInterfaceConfig',
    ]);
    if (parentObject !== undefined && fromPscInterfaceConfig != null) {
        common.setValueByPath(parentObject, ['pscInterfaceConfig'], fromPscInterfaceConfig);
    }
    const fromEncryptionSpec = common.getValueByPath(fromObject, [
        'encryptionSpec',
    ]);
    if (parentObject !== undefined && fromEncryptionSpec != null) {
        common.setValueByPath(parentObject, ['encryptionSpec'], fromEncryptionSpec);
    }
    const fromLabels = common.getValueByPath(fromObject, ['labels']);
    if (parentObject !== undefined && fromLabels != null) {
        common.setValueByPath(parentObject, ['labels'], fromLabels);
    }
    const fromSourcePackages = common.getValueByPath(fromObject, [
        'sourcePackages',
    ]);
    if (parentObject !== undefined && fromSourcePackages != null) {
        common.setValueByPath(parentObject, ['sourcePackages'], fromSourcePackages);
    }
    const fromEntrypointModule = common.getValueByPath(fromObject, [
        'entrypointModule',
    ]);
    if (parentObject !== undefined && fromEntrypointModule != null) {
        common.setValueByPath(parentObject, ['entrypointModule'], fromEntrypointModule);
    }
    const fromEntrypointObject = common.getValueByPath(fromObject, [
        'entrypointObject',
    ]);
    if (parentObject !== undefined && fromEntrypointObject != null) {
        common.setValueByPath(parentObject, ['entrypointObject'], fromEntrypointObject);
    }
    const fromRequirementsFile = common.getValueByPath(fromObject, [
        'requirementsFile',
    ]);
    if (parentObject !== undefined && fromRequirementsFile != null) {
        common.setValueByPath(parentObject, ['requirementsFile'], fromRequirementsFile);
    }
    const fromAgentFramework = common.getValueByPath(fromObject, [
        'agentFramework',
    ]);
    if (parentObject !== undefined && fromAgentFramework != null) {
        common.setValueByPath(parentObject, ['agentFramework'], fromAgentFramework);
    }
    const fromPythonVersion = common.getValueByPath(fromObject, [
        'pythonVersion',
    ]);
    if (parentObject !== undefined && fromPythonVersion != null) {
        common.setValueByPath(parentObject, ['pythonVersion'], fromPythonVersion);
    }
    const fromAgentGatewayConfig = common.getValueByPath(fromObject, [
        'agentGatewayConfig',
    ]);
    if (parentObject !== undefined && fromAgentGatewayConfig != null) {
        common.setValueByPath(parentObject, ['agentGatewayConfig'], fromAgentGatewayConfig);
    }
    return toObject;
}
exports.createAgentEngineConfigToVertex = createAgentEngineConfigToVertex;
function createAgentEngineRequestParametersToVertex(fromObject) {
    const toObject = {};
    const fromConfig = common.getValueByPath(fromObject, ['config']);
    if (fromConfig != null) {
        createAgentEngineConfigToVertex(fromConfig, toObject);
    }
    return toObject;
}
exports.createAgentEngineRequestParametersToVertex = createAgentEngineRequestParametersToVertex;
function deleteAgentEngineRequestParametersToVertex(fromObject) {
    const toObject = {};
    const fromName = common.getValueByPath(fromObject, ['name']);
    if (fromName != null) {
        common.setValueByPath(toObject, ['_url', 'name'], fromName);
    }
    const fromForce = common.getValueByPath(fromObject, ['force']);
    if (fromForce != null) {
        common.setValueByPath(toObject, ['force'], fromForce);
    }
    return toObject;
}
exports.deleteAgentEngineRequestParametersToVertex = deleteAgentEngineRequestParametersToVertex;
function getAgentEngineOperationParametersToVertex(fromObject) {
    const toObject = {};
    const fromOperationName = common.getValueByPath(fromObject, [
        'operationName',
    ]);
    if (fromOperationName != null) {
        common.setValueByPath(toObject, ['_url', 'operationName'], fromOperationName);
    }
    return toObject;
}
exports.getAgentEngineOperationParametersToVertex = getAgentEngineOperationParametersToVertex;
function getAgentEngineRequestParametersToVertex(fromObject) {
    const toObject = {};
    const fromName = common.getValueByPath(fromObject, ['name']);
    if (fromName != null) {
        common.setValueByPath(toObject, ['_url', 'name'], fromName);
    }
    return toObject;
}
exports.getAgentEngineRequestParametersToVertex = getAgentEngineRequestParametersToVertex;
function listAgentEngineConfigToVertex(fromObject, parentObject) {
    const toObject = {};
    const fromPageSize = common.getValueByPath(fromObject, ['pageSize']);
    if (parentObject !== undefined && fromPageSize != null) {
        common.setValueByPath(parentObject, ['_query', 'pageSize'], fromPageSize);
    }
    const fromPageToken = common.getValueByPath(fromObject, ['pageToken']);
    if (parentObject !== undefined && fromPageToken != null) {
        common.setValueByPath(parentObject, ['_query', 'pageToken'], fromPageToken);
    }
    const fromFilter = common.getValueByPath(fromObject, ['filter']);
    if (parentObject !== undefined && fromFilter != null) {
        common.setValueByPath(parentObject, ['_query', 'filter'], fromFilter);
    }
    return toObject;
}
exports.listAgentEngineConfigToVertex = listAgentEngineConfigToVertex;
function listAgentEngineRequestParametersToVertex(fromObject) {
    const toObject = {};
    const fromConfig = common.getValueByPath(fromObject, ['config']);
    if (fromConfig != null) {
        listAgentEngineConfigToVertex(fromConfig, toObject);
    }
    return toObject;
}
exports.listAgentEngineRequestParametersToVertex = listAgentEngineRequestParametersToVertex;
function listReasoningEnginesResponseFromVertex(fromObject) {
    const toObject = {};
    const fromSdkHttpResponse = common.getValueByPath(fromObject, [
        'sdkHttpResponse',
    ]);
    if (fromSdkHttpResponse != null) {
        common.setValueByPath(toObject, ['sdkHttpResponse'], fromSdkHttpResponse);
    }
    const fromNextPageToken = common.getValueByPath(fromObject, [
        'nextPageToken',
    ]);
    if (fromNextPageToken != null) {
        common.setValueByPath(toObject, ['nextPageToken'], fromNextPageToken);
    }
    const fromReasoningEngines = common.getValueByPath(fromObject, [
        'reasoningEngines',
    ]);
    if (fromReasoningEngines != null) {
        let transformedList = fromReasoningEngines;
        if (Array.isArray(transformedList)) {
            transformedList = transformedList.map((item) => {
                return reasoningEngineFromVertex(item);
            });
        }
        common.setValueByPath(toObject, ['reasoningEngines'], transformedList);
    }
    return toObject;
}
exports.listReasoningEnginesResponseFromVertex = listReasoningEnginesResponseFromVertex;
function queryAgentEngineConfigToVertex(fromObject, parentObject) {
    const toObject = {};
    const fromClassMethod = common.getValueByPath(fromObject, ['classMethod']);
    if (parentObject !== undefined && fromClassMethod != null) {
        common.setValueByPath(parentObject, ['classMethod'], fromClassMethod);
    }
    const fromInput = common.getValueByPath(fromObject, ['input']);
    if (parentObject !== undefined && fromInput != null) {
        common.setValueByPath(parentObject, ['input'], fromInput);
    }
    const fromIncludeAllFields = common.getValueByPath(fromObject, [
        'includeAllFields',
    ]);
    if (fromIncludeAllFields != null) {
        common.setValueByPath(toObject, ['includeAllFields'], fromIncludeAllFields);
    }
    return toObject;
}
exports.queryAgentEngineConfigToVertex = queryAgentEngineConfigToVertex;
function queryAgentEngineRequestParametersToVertex(fromObject) {
    const toObject = {};
    const fromName = common.getValueByPath(fromObject, ['name']);
    if (fromName != null) {
        common.setValueByPath(toObject, ['_url', 'name'], fromName);
    }
    const fromConfig = common.getValueByPath(fromObject, ['config']);
    if (fromConfig != null) {
        queryAgentEngineConfigToVertex(fromConfig, toObject);
    }
    return toObject;
}
exports.queryAgentEngineRequestParametersToVertex = queryAgentEngineRequestParametersToVertex;
function reasoningEngineContextSpecFromVertex(fromObject) {
    const toObject = {};
    const fromMemoryBankConfig = common.getValueByPath(fromObject, [
        'memoryBankConfig',
    ]);
    if (fromMemoryBankConfig != null) {
        common.setValueByPath(toObject, ['memoryBankConfig'], reasoningEngineContextSpecMemoryBankConfigFromVertex(fromMemoryBankConfig));
    }
    return toObject;
}
exports.reasoningEngineContextSpecFromVertex = reasoningEngineContextSpecFromVertex;
function reasoningEngineContextSpecMemoryBankConfigFromVertex(fromObject) {
    const toObject = {};
    const fromCustomizationConfigs = common.getValueByPath(fromObject, [
        'customizationConfigs',
    ]);
    if (fromCustomizationConfigs != null) {
        let transformedList = fromCustomizationConfigs;
        if (Array.isArray(transformedList)) {
            transformedList = transformedList.map((item) => {
                return item;
            });
        }
        common.setValueByPath(toObject, ['customizationConfigs'], transformedList);
    }
    const fromDisableMemoryRevisions = common.getValueByPath(fromObject, [
        'disableMemoryRevisions',
    ]);
    if (fromDisableMemoryRevisions != null) {
        common.setValueByPath(toObject, ['disableMemoryRevisions'], fromDisableMemoryRevisions);
    }
    const fromGenerationConfig = common.getValueByPath(fromObject, [
        'generationConfig',
    ]);
    if (fromGenerationConfig != null) {
        common.setValueByPath(toObject, ['generationConfig'], fromGenerationConfig);
    }
    const fromSimilaritySearchConfig = common.getValueByPath(fromObject, [
        'similaritySearchConfig',
    ]);
    if (fromSimilaritySearchConfig != null) {
        common.setValueByPath(toObject, ['similaritySearchConfig'], fromSimilaritySearchConfig);
    }
    const fromTtlConfig = common.getValueByPath(fromObject, ['ttlConfig']);
    if (fromTtlConfig != null) {
        common.setValueByPath(toObject, ['ttlConfig'], fromTtlConfig);
    }
    const fromStructuredMemoryConfigs = common.getValueByPath(fromObject, [
        'structuredMemoryConfigs',
    ]);
    if (fromStructuredMemoryConfigs != null) {
        let transformedList = fromStructuredMemoryConfigs;
        if (Array.isArray(transformedList)) {
            transformedList = transformedList.map((item) => {
                return structuredMemoryConfigFromVertex(item);
            });
        }
        common.setValueByPath(toObject, ['structuredMemoryConfigs'], transformedList);
    }
    return toObject;
}
exports.reasoningEngineContextSpecMemoryBankConfigFromVertex = reasoningEngineContextSpecMemoryBankConfigFromVertex;
function reasoningEngineContextSpecMemoryBankConfigToVertex(fromObject) {
    const toObject = {};
    const fromCustomizationConfigs = common.getValueByPath(fromObject, [
        'customizationConfigs',
    ]);
    if (fromCustomizationConfigs != null) {
        let transformedList = fromCustomizationConfigs;
        if (Array.isArray(transformedList)) {
            transformedList = transformedList.map((item) => {
                return item;
            });
        }
        common.setValueByPath(toObject, ['customizationConfigs'], transformedList);
    }
    const fromDisableMemoryRevisions = common.getValueByPath(fromObject, [
        'disableMemoryRevisions',
    ]);
    if (fromDisableMemoryRevisions != null) {
        common.setValueByPath(toObject, ['disableMemoryRevisions'], fromDisableMemoryRevisions);
    }
    const fromGenerationConfig = common.getValueByPath(fromObject, [
        'generationConfig',
    ]);
    if (fromGenerationConfig != null) {
        common.setValueByPath(toObject, ['generationConfig'], fromGenerationConfig);
    }
    const fromSimilaritySearchConfig = common.getValueByPath(fromObject, [
        'similaritySearchConfig',
    ]);
    if (fromSimilaritySearchConfig != null) {
        common.setValueByPath(toObject, ['similaritySearchConfig'], fromSimilaritySearchConfig);
    }
    const fromTtlConfig = common.getValueByPath(fromObject, ['ttlConfig']);
    if (fromTtlConfig != null) {
        common.setValueByPath(toObject, ['ttlConfig'], fromTtlConfig);
    }
    const fromStructuredMemoryConfigs = common.getValueByPath(fromObject, [
        'structuredMemoryConfigs',
    ]);
    if (fromStructuredMemoryConfigs != null) {
        let transformedList = fromStructuredMemoryConfigs;
        if (Array.isArray(transformedList)) {
            transformedList = transformedList.map((item) => {
                return structuredMemoryConfigToVertex(item);
            });
        }
        common.setValueByPath(toObject, ['structuredMemoryConfigs'], transformedList);
    }
    return toObject;
}
exports.reasoningEngineContextSpecMemoryBankConfigToVertex = reasoningEngineContextSpecMemoryBankConfigToVertex;
function reasoningEngineContextSpecToVertex(fromObject) {
    const toObject = {};
    const fromMemoryBankConfig = common.getValueByPath(fromObject, [
        'memoryBankConfig',
    ]);
    if (fromMemoryBankConfig != null) {
        common.setValueByPath(toObject, ['memoryBankConfig'], reasoningEngineContextSpecMemoryBankConfigToVertex(fromMemoryBankConfig));
    }
    return toObject;
}
exports.reasoningEngineContextSpecToVertex = reasoningEngineContextSpecToVertex;
function reasoningEngineFromVertex(fromObject) {
    const toObject = {};
    const fromEncryptionSpec = common.getValueByPath(fromObject, [
        'encryptionSpec',
    ]);
    if (fromEncryptionSpec != null) {
        common.setValueByPath(toObject, ['encryptionSpec'], fromEncryptionSpec);
    }
    const fromContextSpec = common.getValueByPath(fromObject, ['contextSpec']);
    if (fromContextSpec != null) {
        common.setValueByPath(toObject, ['contextSpec'], reasoningEngineContextSpecFromVertex(fromContextSpec));
    }
    const fromCreateTime = common.getValueByPath(fromObject, ['createTime']);
    if (fromCreateTime != null) {
        common.setValueByPath(toObject, ['createTime'], fromCreateTime);
    }
    const fromDescription = common.getValueByPath(fromObject, ['description']);
    if (fromDescription != null) {
        common.setValueByPath(toObject, ['description'], fromDescription);
    }
    const fromDisplayName = common.getValueByPath(fromObject, ['displayName']);
    if (fromDisplayName != null) {
        common.setValueByPath(toObject, ['displayName'], fromDisplayName);
    }
    const fromEtag = common.getValueByPath(fromObject, ['etag']);
    if (fromEtag != null) {
        common.setValueByPath(toObject, ['etag'], fromEtag);
    }
    const fromLabels = common.getValueByPath(fromObject, ['labels']);
    if (fromLabels != null) {
        common.setValueByPath(toObject, ['labels'], fromLabels);
    }
    const fromName = common.getValueByPath(fromObject, ['name']);
    if (fromName != null) {
        common.setValueByPath(toObject, ['name'], fromName);
    }
    const fromSpec = common.getValueByPath(fromObject, ['spec']);
    if (fromSpec != null) {
        common.setValueByPath(toObject, ['spec'], fromSpec);
    }
    const fromUpdateTime = common.getValueByPath(fromObject, ['updateTime']);
    if (fromUpdateTime != null) {
        common.setValueByPath(toObject, ['updateTime'], fromUpdateTime);
    }
    const fromTrafficConfig = common.getValueByPath(fromObject, [
        'trafficConfig',
    ]);
    if (fromTrafficConfig != null) {
        common.setValueByPath(toObject, ['trafficConfig'], fromTrafficConfig);
    }
    return toObject;
}
exports.reasoningEngineFromVertex = reasoningEngineFromVertex;
function structuredMemoryConfigFromVertex(fromObject) {
    const toObject = {};
    const fromSchemaConfigs = common.getValueByPath(fromObject, [
        'schemaConfigs',
    ]);
    if (fromSchemaConfigs != null) {
        let transformedList = fromSchemaConfigs;
        if (Array.isArray(transformedList)) {
            transformedList = transformedList.map((item) => {
                return structuredMemorySchemaConfigFromVertex(item);
            });
        }
        common.setValueByPath(toObject, ['schemaConfigs'], transformedList);
    }
    const fromScopeKeys = common.getValueByPath(fromObject, ['scopeKeys']);
    if (fromScopeKeys != null) {
        common.setValueByPath(toObject, ['scopeKeys'], fromScopeKeys);
    }
    return toObject;
}
exports.structuredMemoryConfigFromVertex = structuredMemoryConfigFromVertex;
function structuredMemoryConfigToVertex(fromObject) {
    const toObject = {};
    const fromSchemaConfigs = common.getValueByPath(fromObject, [
        'schemaConfigs',
    ]);
    if (fromSchemaConfigs != null) {
        let transformedList = fromSchemaConfigs;
        if (Array.isArray(transformedList)) {
            transformedList = transformedList.map((item) => {
                return structuredMemorySchemaConfigToVertex(item);
            });
        }
        common.setValueByPath(toObject, ['schemaConfigs'], transformedList);
    }
    const fromScopeKeys = common.getValueByPath(fromObject, ['scopeKeys']);
    if (fromScopeKeys != null) {
        common.setValueByPath(toObject, ['scopeKeys'], fromScopeKeys);
    }
    return toObject;
}
exports.structuredMemoryConfigToVertex = structuredMemoryConfigToVertex;
function structuredMemorySchemaConfigFromVertex(fromObject) {
    const toObject = {};
    const fromMemorySchema = common.getValueByPath(fromObject, ['schema']);
    if (fromMemorySchema != null) {
        common.setValueByPath(toObject, ['memorySchema'], fromMemorySchema);
    }
    const fromId = common.getValueByPath(fromObject, ['id']);
    if (fromId != null) {
        common.setValueByPath(toObject, ['id'], fromId);
    }
    const fromMemoryType = common.getValueByPath(fromObject, ['memoryType']);
    if (fromMemoryType != null) {
        common.setValueByPath(toObject, ['memoryType'], fromMemoryType);
    }
    return toObject;
}
exports.structuredMemorySchemaConfigFromVertex = structuredMemorySchemaConfigFromVertex;
function structuredMemorySchemaConfigToVertex(fromObject) {
    const toObject = {};
    const fromMemorySchema = common.getValueByPath(fromObject, ['memorySchema']);
    if (fromMemorySchema != null) {
        common.setValueByPath(toObject, ['schema'], fromMemorySchema);
    }
    const fromId = common.getValueByPath(fromObject, ['id']);
    if (fromId != null) {
        common.setValueByPath(toObject, ['id'], fromId);
    }
    const fromMemoryType = common.getValueByPath(fromObject, ['memoryType']);
    if (fromMemoryType != null) {
        common.setValueByPath(toObject, ['memoryType'], fromMemoryType);
    }
    return toObject;
}
exports.structuredMemorySchemaConfigToVertex = structuredMemorySchemaConfigToVertex;
function updateAgentEngineConfigToVertex(fromObject, parentObject) {
    const toObject = {};
    const fromDisplayName = common.getValueByPath(fromObject, ['displayName']);
    if (parentObject !== undefined && fromDisplayName != null) {
        common.setValueByPath(parentObject, ['displayName'], fromDisplayName);
    }
    const fromDescription = common.getValueByPath(fromObject, ['description']);
    if (parentObject !== undefined && fromDescription != null) {
        common.setValueByPath(parentObject, ['description'], fromDescription);
    }
    const fromSpec = common.getValueByPath(fromObject, ['spec']);
    if (parentObject !== undefined && fromSpec != null) {
        common.setValueByPath(parentObject, ['spec'], fromSpec);
    }
    const fromContextSpec = common.getValueByPath(fromObject, ['contextSpec']);
    if (parentObject !== undefined && fromContextSpec != null) {
        common.setValueByPath(parentObject, ['contextSpec'], reasoningEngineContextSpecToVertex(fromContextSpec));
    }
    const fromPscInterfaceConfig = common.getValueByPath(fromObject, [
        'pscInterfaceConfig',
    ]);
    if (parentObject !== undefined && fromPscInterfaceConfig != null) {
        common.setValueByPath(parentObject, ['pscInterfaceConfig'], fromPscInterfaceConfig);
    }
    const fromEncryptionSpec = common.getValueByPath(fromObject, [
        'encryptionSpec',
    ]);
    if (parentObject !== undefined && fromEncryptionSpec != null) {
        common.setValueByPath(parentObject, ['encryptionSpec'], fromEncryptionSpec);
    }
    const fromLabels = common.getValueByPath(fromObject, ['labels']);
    if (parentObject !== undefined && fromLabels != null) {
        common.setValueByPath(parentObject, ['labels'], fromLabels);
    }
    const fromSourcePackages = common.getValueByPath(fromObject, [
        'sourcePackages',
    ]);
    if (parentObject !== undefined && fromSourcePackages != null) {
        common.setValueByPath(parentObject, ['sourcePackages'], fromSourcePackages);
    }
    const fromEntrypointModule = common.getValueByPath(fromObject, [
        'entrypointModule',
    ]);
    if (parentObject !== undefined && fromEntrypointModule != null) {
        common.setValueByPath(parentObject, ['entrypointModule'], fromEntrypointModule);
    }
    const fromEntrypointObject = common.getValueByPath(fromObject, [
        'entrypointObject',
    ]);
    if (parentObject !== undefined && fromEntrypointObject != null) {
        common.setValueByPath(parentObject, ['entrypointObject'], fromEntrypointObject);
    }
    const fromRequirementsFile = common.getValueByPath(fromObject, [
        'requirementsFile',
    ]);
    if (parentObject !== undefined && fromRequirementsFile != null) {
        common.setValueByPath(parentObject, ['requirementsFile'], fromRequirementsFile);
    }
    const fromAgentFramework = common.getValueByPath(fromObject, [
        'agentFramework',
    ]);
    if (parentObject !== undefined && fromAgentFramework != null) {
        common.setValueByPath(parentObject, ['agentFramework'], fromAgentFramework);
    }
    const fromPythonVersion = common.getValueByPath(fromObject, [
        'pythonVersion',
    ]);
    if (parentObject !== undefined && fromPythonVersion != null) {
        common.setValueByPath(parentObject, ['pythonVersion'], fromPythonVersion);
    }
    const fromAgentGatewayConfig = common.getValueByPath(fromObject, [
        'agentGatewayConfig',
    ]);
    if (parentObject !== undefined && fromAgentGatewayConfig != null) {
        common.setValueByPath(parentObject, ['agentGatewayConfig'], fromAgentGatewayConfig);
    }
    const fromUpdateMask = common.getValueByPath(fromObject, ['updateMask']);
    if (parentObject !== undefined && fromUpdateMask != null) {
        common.setValueByPath(parentObject, ['_query', 'updateMask'], fromUpdateMask);
    }
    return toObject;
}
exports.updateAgentEngineConfigToVertex = updateAgentEngineConfigToVertex;
function updateAgentEngineRequestParametersToVertex(fromObject) {
    const toObject = {};
    const fromName = common.getValueByPath(fromObject, ['name']);
    if (fromName != null) {
        common.setValueByPath(toObject, ['_url', 'name'], fromName);
    }
    const fromConfig = common.getValueByPath(fromObject, ['config']);
    if (fromConfig != null) {
        updateAgentEngineConfigToVertex(fromConfig, toObject);
    }
    return toObject;
}
exports.updateAgentEngineRequestParametersToVertex = updateAgentEngineRequestParametersToVertex;
