// @google-cloud/vertexai/build/src/genai/converters/_memories_converters.js
"use strict";
/**
 * @license
 * Copyright 2025 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.updateAgentEngineMemoryRequestParametersToVertex = exports.updateAgentEngineMemoryConfigToVertex = exports.rollbackAgentEngineMemoryRequestParametersToVertex = exports.retrieveMemoryProfilesRequestParametersToVertex = exports.retrieveAgentEngineMemoriesRequestParametersToVertex = exports.retrieveAgentEngineMemoriesConfigToVertex = exports.purgeAgentEngineMemoriesRequestParametersToVertex = exports.listAgentEngineMemoryRequestParametersToVertex = exports.listAgentEngineMemoryConfigToVertex = exports.ingestEventsRequestParametersToVertex = exports.ingestEventsConfigToVertex = exports.getAgentEngineMemoryRequestParametersToVertex = exports.getAgentEngineMemoryOperationParametersToVertex = exports.getAgentEngineGenerateMemoriesOperationParametersToVertex = exports.generateAgentEngineMemoriesRequestParametersToVertex = exports.generateAgentEngineMemoriesConfigToVertex = exports.deleteAgentEngineMemoryRequestParametersToVertex = exports.createAgentEngineMemoryRequestParametersToVertex = exports.agentEngineMemoryConfigToVertex = void 0;
// Code generated by the Google Gen AI SDK generator DO NOT EDIT.
const common = require("@google/genai/vertex_internal");
function agentEngineMemoryConfigToVertex(fromObject, parentObject) {
    const toObject = {};
    const fromDisplayName = common.getValueByPath(fromObject, ['displayName']);
    if (parentObject !== undefined && fromDisplayName != null) {
        common.setValueByPath(parentObject, ['displayName'], fromDisplayName);
    }
    const fromDescription = common.getValueByPath(fromObject, ['description']);
    if (parentObject !== undefined && fromDescription != null) {
        common.setValueByPath(parentObject, ['description'], fromDescription);
    }
    const fromTtl = common.getValueByPath(fromObject, ['ttl']);
    if (parentObject !== undefined && fromTtl != null) {
        common.setValueByPath(parentObject, ['ttl'], fromTtl);
    }
    const fromExpireTime = common.getValueByPath(fromObject, ['expireTime']);
    if (parentObject !== undefined && fromExpireTime != null) {
        common.setValueByPath(parentObject, ['expireTime'], fromExpireTime);
    }
    const fromRevisionExpireTime = common.getValueByPath(fromObject, [
        'revisionExpireTime',
    ]);
    if (parentObject !== undefined && fromRevisionExpireTime != null) {
        common.setValueByPath(parentObject, ['revisionExpireTime'], fromRevisionExpireTime);
    }
    const fromRevisionTtl = common.getValueByPath(fromObject, ['revisionTtl']);
    if (parentObject !== undefined && fromRevisionTtl != null) {
        common.setValueByPath(parentObject, ['revisionTtl'], fromRevisionTtl);
    }
    const fromDisableMemoryRevisions = common.getValueByPath(fromObject, [
        'disableMemoryRevisions',
    ]);
    if (parentObject !== undefined && fromDisableMemoryRevisions != null) {
        common.setValueByPath(parentObject, ['disableMemoryRevisions'], fromDisableMemoryRevisions);
    }
    const fromTopics = common.getValueByPath(fromObject, ['topics']);
    if (parentObject !== undefined && fromTopics != null) {
        let transformedList = fromTopics;
        if (Array.isArray(transformedList)) {
            transformedList = transformedList.map((item) => {
                return item;
            });
        }
        common.setValueByPath(parentObject, ['topics'], transformedList);
    }
    const fromMetadata = common.getValueByPath(fromObject, ['metadata']);
    if (parentObject !== undefined && fromMetadata != null) {
        common.setValueByPath(parentObject, ['metadata'], fromMetadata);
    }
    const fromMemoryId = common.getValueByPath(fromObject, ['memoryId']);
    if (parentObject !== undefined && fromMemoryId != null) {
        common.setValueByPath(parentObject, ['_query', 'memoryId'], fromMemoryId);
    }
    return toObject;
}
exports.agentEngineMemoryConfigToVertex = agentEngineMemoryConfigToVertex;
function createAgentEngineMemoryRequestParametersToVertex(fromObject) {
    const toObject = {};
    const fromName = common.getValueByPath(fromObject, ['name']);
    if (fromName != null) {
        common.setValueByPath(toObject, ['_url', 'name'], fromName);
    }
    const fromFact = common.getValueByPath(fromObject, ['fact']);
    if (fromFact != null) {
        common.setValueByPath(toObject, ['fact'], fromFact);
    }
    const fromScope = common.getValueByPath(fromObject, ['scope']);
    if (fromScope != null) {
        common.setValueByPath(toObject, ['scope'], fromScope);
    }
    const fromConfig = common.getValueByPath(fromObject, ['config']);
    if (fromConfig != null) {
        agentEngineMemoryConfigToVertex(fromConfig, toObject);
    }
    return toObject;
}
exports.createAgentEngineMemoryRequestParametersToVertex = createAgentEngineMemoryRequestParametersToVertex;
function deleteAgentEngineMemoryRequestParametersToVertex(fromObject) {
    const toObject = {};
    const fromName = common.getValueByPath(fromObject, ['name']);
    if (fromName != null) {
        common.setValueByPath(toObject, ['_url', 'name'], fromName);
    }
    return toObject;
}
exports.deleteAgentEngineMemoryRequestParametersToVertex = deleteAgentEngineMemoryRequestParametersToVertex;
function generateAgentEngineMemoriesConfigToVertex(fromObject, parentObject) {
    const toObject = {};
    const fromDisableConsolidation = common.getValueByPath(fromObject, [
        'disableConsolidation',
    ]);
    if (parentObject !== undefined && fromDisableConsolidation != null) {
        common.setValueByPath(parentObject, ['disableConsolidation'], fromDisableConsolidation);
    }
    const fromRevisionLabels = common.getValueByPath(fromObject, [
        'revisionLabels',
    ]);
    if (parentObject !== undefined && fromRevisionLabels != null) {
        common.setValueByPath(parentObject, ['revisionLabels'], fromRevisionLabels);
    }
    const fromRevisionExpireTime = common.getValueByPath(fromObject, [
        'revisionExpireTime',
    ]);
    if (parentObject !== undefined && fromRevisionExpireTime != null) {
        common.setValueByPath(parentObject, ['revisionExpireTime'], fromRevisionExpireTime);
    }
    const fromRevisionTtl = common.getValueByPath(fromObject, ['revisionTtl']);
    if (parentObject !== undefined && fromRevisionTtl != null) {
        common.setValueByPath(parentObject, ['revisionTtl'], fromRevisionTtl);
    }
    const fromDisableMemoryRevisions = common.getValueByPath(fromObject, [
        'disableMemoryRevisions',
    ]);
    if (parentObject !== undefined && fromDisableMemoryRevisions != null) {
        common.setValueByPath(parentObject, ['disableMemoryRevisions'], fromDisableMemoryRevisions);
    }
    const fromMetadata = common.getValueByPath(fromObject, ['metadata']);
    if (parentObject !== undefined && fromMetadata != null) {
        common.setValueByPath(parentObject, ['metadata'], fromMetadata);
    }
    const fromMetadataMergeStrategy = common.getValueByPath(fromObject, [
        'metadataMergeStrategy',
    ]);
    if (parentObject !== undefined && fromMetadataMergeStrategy != null) {
        common.setValueByPath(parentObject, ['metadataMergeStrategy'], fromMetadataMergeStrategy);
    }
    const fromAllowedTopics = common.getValueByPath(fromObject, [
        'allowedTopics',
    ]);
    if (parentObject !== undefined && fromAllowedTopics != null) {
        let transformedList = fromAllowedTopics;
        if (Array.isArray(transformedList)) {
            transformedList = transformedList.map((item) => {
                return item;
            });
        }
        common.setValueByPath(parentObject, ['allowedTopics'], transformedList);
    }
    return toObject;
}
exports.generateAgentEngineMemoriesConfigToVertex = generateAgentEngineMemoriesConfigToVertex;
function generateAgentEngineMemoriesRequestParametersToVertex(fromObject) {
    const toObject = {};
    const fromName = common.getValueByPath(fromObject, ['name']);
    if (fromName != null) {
        common.setValueByPath(toObject, ['_url', 'name'], fromName);
    }
    const fromVertexSessionSource = common.getValueByPath(fromObject, [
        'vertexSessionSource',
    ]);
    if (fromVertexSessionSource != null) {
        common.setValueByPath(toObject, ['vertexSessionSource'], fromVertexSessionSource);
    }
    const fromDirectContentsSource = common.getValueByPath(fromObject, [
        'directContentsSource',
    ]);
    if (fromDirectContentsSource != null) {
        common.setValueByPath(toObject, ['directContentsSource'], fromDirectContentsSource);
    }
    const fromDirectMemoriesSource = common.getValueByPath(fromObject, [
        'directMemoriesSource',
    ]);
    if (fromDirectMemoriesSource != null) {
        common.setValueByPath(toObject, ['directMemoriesSource'], fromDirectMemoriesSource);
    }
    const fromScope = common.getValueByPath(fromObject, ['scope']);
    if (fromScope != null) {
        common.setValueByPath(toObject, ['scope'], fromScope);
    }
    const fromConfig = common.getValueByPath(fromObject, ['config']);
    if (fromConfig != null) {
        generateAgentEngineMemoriesConfigToVertex(fromConfig, toObject);
    }
    return toObject;
}
exports.generateAgentEngineMemoriesRequestParametersToVertex = generateAgentEngineMemoriesRequestParametersToVertex;
function getAgentEngineGenerateMemoriesOperationParametersToVertex(fromObject) {
    const toObject = {};
    const fromOperationName = common.getValueByPath(fromObject, [
        'operationName',
    ]);
    if (fromOperationName != null) {
        common.setValueByPath(toObject, ['_url', 'operationName'], fromOperationName);
    }
    return toObject;
}
exports.getAgentEngineGenerateMemoriesOperationParametersToVertex = getAgentEngineGenerateMemoriesOperationParametersToVertex;
function getAgentEngineMemoryOperationParametersToVertex(fromObject) {
    const toObject = {};
    const fromOperationName = common.getValueByPath(fromObject, [
        'operationName',
    ]);
    if (fromOperationName != null) {
        common.setValueByPath(toObject, ['_url', 'operationName'], fromOperationName);
    }
    return toObject;
}
exports.getAgentEngineMemoryOperationParametersToVertex = getAgentEngineMemoryOperationParametersToVertex;
function getAgentEngineMemoryRequestParametersToVertex(fromObject) {
    const toObject = {};
    const fromName = common.getValueByPath(fromObject, ['name']);
    if (fromName != null) {
        common.setValueByPath(toObject, ['_url', 'name'], fromName);
    }
    return toObject;
}
exports.getAgentEngineMemoryRequestParametersToVertex = getAgentEngineMemoryRequestParametersToVertex;
function ingestEventsConfigToVertex(fromObject, parentObject) {
    const toObject = {};
    const fromForceFlush = common.getValueByPath(fromObject, ['forceFlush']);
    if (parentObject !== undefined && fromForceFlush != null) {
        common.setValueByPath(parentObject, ['forceFlush'], fromForceFlush);
    }
    return toObject;
}
exports.ingestEventsConfigToVertex = ingestEventsConfigToVertex;
function ingestEventsRequestParametersToVertex(fromObject) {
    const toObject = {};
    const fromName = common.getValueByPath(fromObject, ['name']);
    if (fromName != null) {
        common.setValueByPath(toObject, ['_url', 'name'], fromName);
    }
    const fromStreamId = common.getValueByPath(fromObject, ['streamId']);
    if (fromStreamId != null) {
        common.setValueByPath(toObject, ['streamId'], fromStreamId);
    }
    const fromDirectContentsSource = common.getValueByPath(fromObject, [
        'directContentsSource',
    ]);
    if (fromDirectContentsSource != null) {
        common.setValueByPath(toObject, ['directContentsSource'], fromDirectContentsSource);
    }
    const fromScope = common.getValueByPath(fromObject, ['scope']);
    if (fromScope != null) {
        common.setValueByPath(toObject, ['scope'], fromScope);
    }
    const fromGenerationTriggerConfig = common.getValueByPath(fromObject, [
        'generationTriggerConfig',
    ]);
    if (fromGenerationTriggerConfig != null) {
        common.setValueByPath(toObject, ['generationTriggerConfig'], fromGenerationTriggerConfig);
    }
    const fromConfig = common.getValueByPath(fromObject, ['config']);
    if (fromConfig != null) {
        ingestEventsConfigToVertex(fromConfig, toObject);
    }
    return toObject;
}
exports.ingestEventsRequestParametersToVertex = ingestEventsRequestParametersToVertex;
function listAgentEngineMemoryConfigToVertex(fromObject, parentObject) {
    const toObject = {};
    const fromPageSize = common.getValueByPath(fromObject, ['pageSize']);
    if (parentObject !== undefined && fromPageSize != null) {
        common.setValueByPath(parentObject, ['_query', 'pageSize'], fromPageSize);
    }
    const fromPageToken = common.getValueByPath(fromObject, ['pageToken']);
    if (parentObject !== undefined && fromPageToken != null) {
        common.setValueByPath(parentObject, ['_query', 'pageToken'], fromPageToken);
    }
    const fromFilter = common.getValueByPath(fromObject, ['filter']);
    if (parentObject !== undefined && fromFilter != null) {
        common.setValueByPath(parentObject, ['_query', 'filter'], fromFilter);
    }
    const fromOrderBy = common.getValueByPath(fromObject, ['orderBy']);
    if (parentObject !== undefined && fromOrderBy != null) {
        common.setValueByPath(parentObject, ['_query', 'orderBy'], fromOrderBy);
    }
    return toObject;
}
exports.listAgentEngineMemoryConfigToVertex = listAgentEngineMemoryConfigToVertex;
function listAgentEngineMemoryRequestParametersToVertex(fromObject) {
    const toObject = {};
    const fromName = common.getValueByPath(fromObject, ['name']);
    if (fromName != null) {
        common.setValueByPath(toObject, ['_url', 'name'], fromName);
    }
    const fromConfig = common.getValueByPath(fromObject, ['config']);
    if (fromConfig != null) {
        listAgentEngineMemoryConfigToVertex(fromConfig, toObject);
    }
    return toObject;
}
exports.listAgentEngineMemoryRequestParametersToVertex = listAgentEngineMemoryRequestParametersToVertex;
function purgeAgentEngineMemoriesRequestParametersToVertex(fromObject) {
    const toObject = {};
    const fromName = common.getValueByPath(fromObject, ['name']);
    if (fromName != null) {
        common.setValueByPath(toObject, ['_url', 'name'], fromName);
    }
    const fromFilter = common.getValueByPath(fromObject, ['filter']);
    if (fromFilter != null) {
        common.setValueByPath(toObject, ['filter'], fromFilter);
    }
    const fromFilterGroups = common.getValueByPath(fromObject, ['filterGroups']);
    if (fromFilterGroups != null) {
        let transformedList = fromFilterGroups;
        if (Array.isArray(transformedList)) {
            transformedList = transformedList.map((item) => {
                return item;
            });
        }
        common.setValueByPath(toObject, ['filterGroups'], transformedList);
    }
    const fromForce = common.getValueByPath(fromObject, ['force']);
    if (fromForce != null) {
        common.setValueByPath(toObject, ['force'], fromForce);
    }
    return toObject;
}
exports.purgeAgentEngineMemoriesRequestParametersToVertex = purgeAgentEngineMemoriesRequestParametersToVertex;
function retrieveAgentEngineMemoriesConfigToVertex(fromObject, parentObject) {
    const toObject = {};
    const fromFilter = common.getValueByPath(fromObject, ['filter']);
    if (parentObject !== undefined && fromFilter != null) {
        common.setValueByPath(parentObject, ['filter'], fromFilter);
    }
    const fromFilterGroups = common.getValueByPath(fromObject, ['filterGroups']);
    if (parentObject !== undefined && fromFilterGroups != null) {
        let transformedList = fromFilterGroups;
        if (Array.isArray(transformedList)) {
            transformedList = transformedList.map((item) => {
                return item;
            });
        }
        common.setValueByPath(parentObject, ['filterGroups'], transformedList);
    }
    const fromMemoryTypes = common.getValueByPath(fromObject, ['memoryTypes']);
    if (parentObject !== undefined && fromMemoryTypes != null) {
        common.setValueByPath(parentObject, ['memoryTypes'], fromMemoryTypes);
    }
    return toObject;
}
exports.retrieveAgentEngineMemoriesConfigToVertex = retrieveAgentEngineMemoriesConfigToVertex;
function retrieveAgentEngineMemoriesRequestParametersToVertex(fromObject) {
    const toObject = {};
    const fromName = common.getValueByPath(fromObject, ['name']);
    if (fromName != null) {
        common.setValueByPath(toObject, ['_url', 'name'], fromName);
    }
    const fromScope = common.getValueByPath(fromObject, ['scope']);
    if (fromScope != null) {
        common.setValueByPath(toObject, ['scope'], fromScope);
    }
    const fromSimilaritySearchParams = common.getValueByPath(fromObject, [
        'similaritySearchParams',
    ]);
    if (fromSimilaritySearchParams != null) {
        common.setValueByPath(toObject, ['similaritySearchParams'], fromSimilaritySearchParams);
    }
    const fromSimpleRetrievalParams = common.getValueByPath(fromObject, [
        'simpleRetrievalParams',
    ]);
    if (fromSimpleRetrievalParams != null) {
        common.setValueByPath(toObject, ['simpleRetrievalParams'], fromSimpleRetrievalParams);
    }
    const fromConfig = common.getValueByPath(fromObject, ['config']);
    if (fromConfig != null) {
        retrieveAgentEngineMemoriesConfigToVertex(fromConfig, toObject);
    }
    return toObject;
}
exports.retrieveAgentEngineMemoriesRequestParametersToVertex = retrieveAgentEngineMemoriesRequestParametersToVertex;
function retrieveMemoryProfilesRequestParametersToVertex(fromObject) {
    const toObject = {};
    const fromName = common.getValueByPath(fromObject, ['name']);
    if (fromName != null) {
        common.setValueByPath(toObject, ['_url', 'name'], fromName);
    }
    const fromScope = common.getValueByPath(fromObject, ['scope']);
    if (fromScope != null) {
        common.setValueByPath(toObject, ['scope'], fromScope);
    }
    return toObject;
}
exports.retrieveMemoryProfilesRequestParametersToVertex = retrieveMemoryProfilesRequestParametersToVertex;
function rollbackAgentEngineMemoryRequestParametersToVertex(fromObject) {
    const toObject = {};
    const fromName = common.getValueByPath(fromObject, ['name']);
    if (fromName != null) {
        common.setValueByPath(toObject, ['_url', 'name'], fromName);
    }
    const fromTargetRevisionId = common.getValueByPath(fromObject, [
        'targetRevisionId',
    ]);
    if (fromTargetRevisionId != null) {
        common.setValueByPath(toObject, ['targetRevisionId'], fromTargetRevisionId);
    }
    return toObject;
}
exports.rollbackAgentEngineMemoryRequestParametersToVertex = rollbackAgentEngineMemoryRequestParametersToVertex;
function updateAgentEngineMemoryConfigToVertex(fromObject, parentObject) {
    const toObject = {};
    const fromDisplayName = common.getValueByPath(fromObject, ['displayName']);
    if (parentObject !== undefined && fromDisplayName != null) {
        common.setValueByPath(parentObject, ['displayName'], fromDisplayName);
    }
    const fromDescription = common.getValueByPath(fromObject, ['description']);
    if (parentObject !== undefined && fromDescription != null) {
        common.setValueByPath(parentObject, ['description'], fromDescription);
    }
    const fromTtl = common.getValueByPath(fromObject, ['ttl']);
    if (parentObject !== undefined && fromTtl != null) {
        common.setValueByPath(parentObject, ['ttl'], fromTtl);
    }
    const fromExpireTime = common.getValueByPath(fromObject, ['expireTime']);
    if (parentObject !== undefined && fromExpireTime != null) {
        common.setValueByPath(parentObject, ['expireTime'], fromExpireTime);
    }
    const fromRevisionExpireTime = common.getValueByPath(fromObject, [
        'revisionExpireTime',
    ]);
    if (parentObject !== undefined && fromRevisionExpireTime != null) {
        common.setValueByPath(parentObject, ['revisionExpireTime'], fromRevisionExpireTime);
    }
    const fromRevisionTtl = common.getValueByPath(fromObject, ['revisionTtl']);
    if (parentObject !== undefined && fromRevisionTtl != null) {
        common.setValueByPath(parentObject, ['revisionTtl'], fromRevisionTtl);
    }
    const fromDisableMemoryRevisions = common.getValueByPath(fromObject, [
        'disableMemoryRevisions',
    ]);
    if (parentObject !== undefined && fromDisableMemoryRevisions != null) {
        common.setValueByPath(parentObject, ['disableMemoryRevisions'], fromDisableMemoryRevisions);
    }
    const fromTopics = common.getValueByPath(fromObject, ['topics']);
    if (parentObject !== undefined && fromTopics != null) {
        let transformedList = fromTopics;
        if (Array.isArray(transformedList)) {
            transformedList = transformedList.map((item) => {
                return item;
            });
        }
        common.setValueByPath(parentObject, ['topics'], transformedList);
    }
    const fromMetadata = common.getValueByPath(fromObject, ['metadata']);
    if (parentObject !== undefined && fromMetadata != null) {
        common.setValueByPath(parentObject, ['metadata'], fromMetadata);
    }
    const fromMemoryId = common.getValueByPath(fromObject, ['memoryId']);
    if (parentObject !== undefined && fromMemoryId != null) {
        common.setValueByPath(parentObject, ['_query', 'memoryId'], fromMemoryId);
    }
    const fromUpdateMask = common.getValueByPath(fromObject, ['updateMask']);
    if (parentObject !== undefined && fromUpdateMask != null) {
        common.setValueByPath(parentObject, ['_query', 'updateMask'], fromUpdateMask);
    }
    return toObject;
}
exports.updateAgentEngineMemoryConfigToVertex = updateAgentEngineMemoryConfigToVertex;
function updateAgentEngineMemoryRequestParametersToVertex(fromObject) {
    const toObject = {};
    const fromName = common.getValueByPath(fromObject, ['name']);
    if (fromName != null) {
        common.setValueByPath(toObject, ['_url', 'name'], fromName);
    }
    const fromFact = common.getValueByPath(fromObject, ['fact']);
    if (fromFact != null) {
        common.setValueByPath(toObject, ['fact'], fromFact);
    }
    const fromScope = common.getValueByPath(fromObject, ['scope']);
    if (fromScope != null) {
        common.setValueByPath(toObject, ['scope'], fromScope);
    }
    const fromConfig = common.getValueByPath(fromObject, ['config']);
    if (fromConfig != null) {
        updateAgentEngineMemoryConfigToVertex(fromConfig, toObject);
    }
    return toObject;
}
exports.updateAgentEngineMemoryRequestParametersToVertex = updateAgentEngineMemoryRequestParametersToVertex;
