// @google-cloud/vertexai/build/src/genai/converters/_sandboxes_converters.js
"use strict";
/**
 * @license
 * Copyright 2025 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.listAgentEngineSandboxesRequestParametersToVertex = exports.listAgentEngineSandboxesConfigToVertex = exports.getAgentEngineSandboxRequestParametersToVertex = exports.getAgentEngineSandboxOperationParametersToVertex = exports.executeCodeAgentEngineSandboxRequestParametersToVertex = exports.deleteAgentEngineSandboxRequestParametersToVertex = exports.createAgentEngineSandboxRequestParametersToVertex = exports.createAgentEngineSandboxConfigToVertex = void 0;
// Code generated by the Google Gen AI SDK generator DO NOT EDIT.
const common = require("@google/genai/vertex_internal");
function createAgentEngineSandboxConfigToVertex(fromObject, parentObject) {
    const toObject = {};
    const fromDisplayName = common.getValueByPath(fromObject, ['displayName']);
    if (parentObject !== undefined && fromDisplayName != null) {
        common.setValueByPath(parentObject, ['displayName'], fromDisplayName);
    }
    const fromDescription = common.getValueByPath(fromObject, ['description']);
    if (parentObject !== undefined && fromDescription != null) {
        common.setValueByPath(parentObject, ['description'], fromDescription);
    }
    const fromTtl = common.getValueByPath(fromObject, ['ttl']);
    if (parentObject !== undefined && fromTtl != null) {
        common.setValueByPath(parentObject, ['ttl'], fromTtl);
    }
    return toObject;
}
exports.createAgentEngineSandboxConfigToVertex = createAgentEngineSandboxConfigToVertex;
function createAgentEngineSandboxRequestParametersToVertex(fromObject) {
    const toObject = {};
    const fromName = common.getValueByPath(fromObject, ['name']);
    if (fromName != null) {
        common.setValueByPath(toObject, ['_url', 'name'], fromName);
    }
    const fromSpec = common.getValueByPath(fromObject, ['spec']);
    if (fromSpec != null) {
        common.setValueByPath(toObject, ['spec'], fromSpec);
    }
    const fromConfig = common.getValueByPath(fromObject, ['config']);
    if (fromConfig != null) {
        createAgentEngineSandboxConfigToVertex(fromConfig, toObject);
    }
    return toObject;
}
exports.createAgentEngineSandboxRequestParametersToVertex = createAgentEngineSandboxRequestParametersToVertex;
function deleteAgentEngineSandboxRequestParametersToVertex(fromObject) {
    const toObject = {};
    const fromName = common.getValueByPath(fromObject, ['name']);
    if (fromName != null) {
        common.setValueByPath(toObject, ['_url', 'name'], fromName);
    }
    return toObject;
}
exports.deleteAgentEngineSandboxRequestParametersToVertex = deleteAgentEngineSandboxRequestParametersToVertex;
function executeCodeAgentEngineSandboxRequestParametersToVertex(fromObject) {
    const toObject = {};
    const fromName = common.getValueByPath(fromObject, ['name']);
    if (fromName != null) {
        common.setValueByPath(toObject, ['_url', 'name'], fromName);
    }
    const fromInputs = common.getValueByPath(fromObject, ['inputs']);
    if (fromInputs != null) {
        let transformedList = fromInputs;
        if (Array.isArray(transformedList)) {
            transformedList = transformedList.map((item) => {
                return item;
            });
        }
        common.setValueByPath(toObject, ['inputs'], transformedList);
    }
    return toObject;
}
exports.executeCodeAgentEngineSandboxRequestParametersToVertex = executeCodeAgentEngineSandboxRequestParametersToVertex;
function getAgentEngineSandboxOperationParametersToVertex(fromObject) {
    const toObject = {};
    const fromOperationName = common.getValueByPath(fromObject, [
        'operationName',
    ]);
    if (fromOperationName != null) {
        common.setValueByPath(toObject, ['_url', 'operationName'], fromOperationName);
    }
    return toObject;
}
exports.getAgentEngineSandboxOperationParametersToVertex = getAgentEngineSandboxOperationParametersToVertex;
function getAgentEngineSandboxRequestParametersToVertex(fromObject) {
    const toObject = {};
    const fromName = common.getValueByPath(fromObject, ['name']);
    if (fromName != null) {
        common.setValueByPath(toObject, ['_url', 'name'], fromName);
    }
    return toObject;
}
exports.getAgentEngineSandboxRequestParametersToVertex = getAgentEngineSandboxRequestParametersToVertex;
function listAgentEngineSandboxesConfigToVertex(fromObject, parentObject) {
    const toObject = {};
    const fromPageSize = common.getValueByPath(fromObject, ['pageSize']);
    if (parentObject !== undefined && fromPageSize != null) {
        common.setValueByPath(parentObject, ['_query', 'pageSize'], fromPageSize);
    }
    const fromPageToken = common.getValueByPath(fromObject, ['pageToken']);
    if (parentObject !== undefined && fromPageToken != null) {
        common.setValueByPath(parentObject, ['_query', 'pageToken'], fromPageToken);
    }
    const fromFilter = common.getValueByPath(fromObject, ['filter']);
    if (parentObject !== undefined && fromFilter != null) {
        common.setValueByPath(parentObject, ['_query', 'filter'], fromFilter);
    }
    return toObject;
}
exports.listAgentEngineSandboxesConfigToVertex = listAgentEngineSandboxesConfigToVertex;
function listAgentEngineSandboxesRequestParametersToVertex(fromObject) {
    const toObject = {};
    const fromName = common.getValueByPath(fromObject, ['name']);
    if (fromName != null) {
        common.setValueByPath(toObject, ['_url', 'name'], fromName);
    }
    const fromConfig = common.getValueByPath(fromObject, ['config']);
    if (fromConfig != null) {
        listAgentEngineSandboxesConfigToVertex(fromConfig, toObject);
    }
    return toObject;
}
exports.listAgentEngineSandboxesRequestParametersToVertex = listAgentEngineSandboxesRequestParametersToVertex;
