// @google-cloud/vertexai/build/src/genai/converters/_sessions_converters.js
"use strict";
/**
 * @license
 * Copyright 2025 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.updateAgentEngineSessionRequestParametersToVertex = exports.updateAgentEngineSessionConfigToVertex = exports.listAgentEngineSessionsRequestParametersToVertex = exports.listAgentEngineSessionsConfigToVertex = exports.getAgentEngineSessionRequestParametersToVertex = exports.getAgentEngineSessionOperationParametersToVertex = exports.deleteAgentEngineSessionRequestParametersToVertex = exports.createAgentEngineSessionRequestParametersToVertex = exports.createAgentEngineSessionConfigToVertex = void 0;
// Code generated by the Google Gen AI SDK generator DO NOT EDIT.
const common = require("@google/genai/vertex_internal");
function createAgentEngineSessionConfigToVertex(fromObject, parentObject) {
    const toObject = {};
    const fromDisplayName = common.getValueByPath(fromObject, ['displayName']);
    if (parentObject !== undefined && fromDisplayName != null) {
        common.setValueByPath(parentObject, ['displayName'], fromDisplayName);
    }
    const fromSessionState = common.getValueByPath(fromObject, ['sessionState']);
    if (parentObject !== undefined && fromSessionState != null) {
        common.setValueByPath(parentObject, ['sessionState'], fromSessionState);
    }
    const fromTtl = common.getValueByPath(fromObject, ['ttl']);
    if (parentObject !== undefined && fromTtl != null) {
        common.setValueByPath(parentObject, ['ttl'], fromTtl);
    }
    const fromExpireTime = common.getValueByPath(fromObject, ['expireTime']);
    if (parentObject !== undefined && fromExpireTime != null) {
        common.setValueByPath(parentObject, ['expireTime'], fromExpireTime);
    }
    const fromLabels = common.getValueByPath(fromObject, ['labels']);
    if (parentObject !== undefined && fromLabels != null) {
        common.setValueByPath(parentObject, ['labels'], fromLabels);
    }
    const fromSessionId = common.getValueByPath(fromObject, ['sessionId']);
    if (parentObject !== undefined && fromSessionId != null) {
        common.setValueByPath(parentObject, ['_query', 'sessionId'], fromSessionId);
    }
    return toObject;
}
exports.createAgentEngineSessionConfigToVertex = createAgentEngineSessionConfigToVertex;
function createAgentEngineSessionRequestParametersToVertex(fromObject) {
    const toObject = {};
    const fromName = common.getValueByPath(fromObject, ['name']);
    if (fromName != null) {
        common.setValueByPath(toObject, ['_url', 'name'], fromName);
    }
    const fromUserId = common.getValueByPath(fromObject, ['userId']);
    if (fromUserId != null) {
        common.setValueByPath(toObject, ['userId'], fromUserId);
    }
    const fromConfig = common.getValueByPath(fromObject, ['config']);
    if (fromConfig != null) {
        createAgentEngineSessionConfigToVertex(fromConfig, toObject);
    }
    return toObject;
}
exports.createAgentEngineSessionRequestParametersToVertex = createAgentEngineSessionRequestParametersToVertex;
function deleteAgentEngineSessionRequestParametersToVertex(fromObject) {
    const toObject = {};
    const fromName = common.getValueByPath(fromObject, ['name']);
    if (fromName != null) {
        common.setValueByPath(toObject, ['_url', 'name'], fromName);
    }
    return toObject;
}
exports.deleteAgentEngineSessionRequestParametersToVertex = deleteAgentEngineSessionRequestParametersToVertex;
function getAgentEngineSessionOperationParametersToVertex(fromObject) {
    const toObject = {};
    const fromOperationName = common.getValueByPath(fromObject, [
        'operationName',
    ]);
    if (fromOperationName != null) {
        common.setValueByPath(toObject, ['_url', 'operationName'], fromOperationName);
    }
    return toObject;
}
exports.getAgentEngineSessionOperationParametersToVertex = getAgentEngineSessionOperationParametersToVertex;
function getAgentEngineSessionRequestParametersToVertex(fromObject) {
    const toObject = {};
    const fromName = common.getValueByPath(fromObject, ['name']);
    if (fromName != null) {
        common.setValueByPath(toObject, ['_url', 'name'], fromName);
    }
    return toObject;
}
exports.getAgentEngineSessionRequestParametersToVertex = getAgentEngineSessionRequestParametersToVertex;
function listAgentEngineSessionsConfigToVertex(fromObject, parentObject) {
    const toObject = {};
    const fromPageSize = common.getValueByPath(fromObject, ['pageSize']);
    if (parentObject !== undefined && fromPageSize != null) {
        common.setValueByPath(parentObject, ['_query', 'pageSize'], fromPageSize);
    }
    const fromPageToken = common.getValueByPath(fromObject, ['pageToken']);
    if (parentObject !== undefined && fromPageToken != null) {
        common.setValueByPath(parentObject, ['_query', 'pageToken'], fromPageToken);
    }
    const fromFilter = common.getValueByPath(fromObject, ['filter']);
    if (parentObject !== undefined && fromFilter != null) {
        common.setValueByPath(parentObject, ['_query', 'filter'], fromFilter);
    }
    return toObject;
}
exports.listAgentEngineSessionsConfigToVertex = listAgentEngineSessionsConfigToVertex;
function listAgentEngineSessionsRequestParametersToVertex(fromObject) {
    const toObject = {};
    const fromName = common.getValueByPath(fromObject, ['name']);
    if (fromName != null) {
        common.setValueByPath(toObject, ['_url', 'name'], fromName);
    }
    const fromConfig = common.getValueByPath(fromObject, ['config']);
    if (fromConfig != null) {
        listAgentEngineSessionsConfigToVertex(fromConfig, toObject);
    }
    return toObject;
}
exports.listAgentEngineSessionsRequestParametersToVertex = listAgentEngineSessionsRequestParametersToVertex;
function updateAgentEngineSessionConfigToVertex(fromObject, parentObject) {
    const toObject = {};
    const fromDisplayName = common.getValueByPath(fromObject, ['displayName']);
    if (parentObject !== undefined && fromDisplayName != null) {
        common.setValueByPath(parentObject, ['displayName'], fromDisplayName);
    }
    const fromSessionState = common.getValueByPath(fromObject, ['sessionState']);
    if (parentObject !== undefined && fromSessionState != null) {
        common.setValueByPath(parentObject, ['sessionState'], fromSessionState);
    }
    const fromTtl = common.getValueByPath(fromObject, ['ttl']);
    if (parentObject !== undefined && fromTtl != null) {
        common.setValueByPath(parentObject, ['ttl'], fromTtl);
    }
    const fromExpireTime = common.getValueByPath(fromObject, ['expireTime']);
    if (parentObject !== undefined && fromExpireTime != null) {
        common.setValueByPath(parentObject, ['expireTime'], fromExpireTime);
    }
    const fromLabels = common.getValueByPath(fromObject, ['labels']);
    if (parentObject !== undefined && fromLabels != null) {
        common.setValueByPath(parentObject, ['labels'], fromLabels);
    }
    const fromSessionId = common.getValueByPath(fromObject, ['sessionId']);
    if (parentObject !== undefined && fromSessionId != null) {
        common.setValueByPath(parentObject, ['_query', 'sessionId'], fromSessionId);
    }
    const fromUpdateMask = common.getValueByPath(fromObject, ['updateMask']);
    if (parentObject !== undefined && fromUpdateMask != null) {
        common.setValueByPath(parentObject, ['_query', 'updateMask'], fromUpdateMask);
    }
    const fromUserId = common.getValueByPath(fromObject, ['userId']);
    if (parentObject !== undefined && fromUserId != null) {
        common.setValueByPath(parentObject, ['userId'], fromUserId);
    }
    return toObject;
}
exports.updateAgentEngineSessionConfigToVertex = updateAgentEngineSessionConfigToVertex;
function updateAgentEngineSessionRequestParametersToVertex(fromObject) {
    const toObject = {};
    const fromName = common.getValueByPath(fromObject, ['name']);
    if (fromName != null) {
        common.setValueByPath(toObject, ['_url', 'name'], fromName);
    }
    const fromConfig = common.getValueByPath(fromObject, ['config']);
    if (fromConfig != null) {
        updateAgentEngineSessionConfigToVertex(fromConfig, toObject);
    }
    return toObject;
}
exports.updateAgentEngineSessionRequestParametersToVertex = updateAgentEngineSessionRequestParametersToVertex;
