// @genkit-ai/core/lib/tracing/realtime-span-processor.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var realtime_span_processor_exports = {};
__export(realtime_span_processor_exports, {
  RealtimeSpanProcessor: () => RealtimeSpanProcessor
});
module.exports = __toCommonJS(realtime_span_processor_exports);
class RealtimeSpanProcessor {
  constructor(exporter) {
    this.exporter = exporter;
  }
  /**
   * Called when a span is started. Exports immediately for real-time updates.
   */
  onStart(span, _parentContext) {
    this.exporter.export([span], () => {
    });
  }
  /**
   * Called when a span ends. Exports again with complete data.
   */
  onEnd(span) {
    this.exporter.export([span], () => {
    });
  }
  /**
   * Forces the exporter to flush any buffered spans.
   */
  async forceFlush() {
    if (this.exporter.forceFlush) {
      return this.exporter.forceFlush();
    }
  }
  /**
   * Shuts down the processor and exporter.
   */
  async shutdown() {
    return this.exporter.shutdown();
  }
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  RealtimeSpanProcessor
});
