// @google-cloud/vertexai/build/src/genai/types/common.js
"use strict";
/**
 * @license
 * Copyright 2025 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.CheckQueryJobResponse = exports.ListAgentEngineSessionEventsResponse = exports.AppendAgentEngineSessionEventResponse = exports.ListReasoningEnginesSessionsResponse = exports.ListAgentEngineSandboxesResponse = exports.ExecuteSandboxEnvironmentResponse = exports.ListAgentEngineMemoryRevisionsResponse = exports.PurgeMemoriesResponse = exports.RetrieveProfilesResponse = exports.RetrieveMemoriesResponse = exports.RetrieveMemoriesResponseRetrievedMemory = exports.ListReasoningEnginesMemoriesResponse = exports.GenerateMemoriesResponse = exports.GenerateMemoriesResponseGeneratedMemory = exports.QueryReasoningEngineResponse = exports.ListReasoningEnginesResponse = exports.GenerateMemoriesResponseGeneratedMemoryAction = exports.MemoryMetadataMergeStrategy = exports.Framework = exports.State = exports.MachineConfig = exports.Language = exports.Operator = exports.MemoryType = exports.ManagedTopicEnum = exports.AgentServerMode = exports.IdentityType = void 0;
/** The identity type to use for the Reasoning Engine. If not specified, the `service_account` field will be used if set, otherwise the default Vertex AI Reasoning Engine Service Agent in the project will be used. */
var IdentityType;
(function (IdentityType) {
    /**
     * Default value. Use a custom service account if the `service_account` field is set, otherwise use the default Vertex AI Reasoning Engine Service Agent in the project. Same behavior as SERVICE_ACCOUNT.
     */
    IdentityType["IDENTITY_TYPE_UNSPECIFIED"] = "IDENTITY_TYPE_UNSPECIFIED";
    /**
     * Use a custom service account if the `service_account` field is set, otherwise use the default Vertex AI Reasoning Engine Service Agent in the project.
     */
    IdentityType["SERVICE_ACCOUNT"] = "SERVICE_ACCOUNT";
    /**
     * Use Agent Identity. The `service_account` field must not be set.
     */
    IdentityType["AGENT_IDENTITY"] = "AGENT_IDENTITY";
})(IdentityType || (exports.IdentityType = IdentityType = {}));
/** The agent server mode. */
var AgentServerMode;
(function (AgentServerMode) {
    /**
     * Unspecified agent server mode. Do not use.
     */
    AgentServerMode["AGENT_SERVER_MODE_UNSPECIFIED"] = "AGENT_SERVER_MODE_UNSPECIFIED";
    /**
     * Stable agent server mode. This mode has everything stable and well-tested features agent engine offers.
     */
    AgentServerMode["STABLE"] = "STABLE";
    /**
     * Experimental agent server mode. This mode contains experimental features.
     */
    AgentServerMode["EXPERIMENTAL"] = "EXPERIMENTAL";
})(AgentServerMode || (exports.AgentServerMode = AgentServerMode = {}));
/** Represents the managed memory topic. */
var ManagedTopicEnum;
(function (ManagedTopicEnum) {
    /**
     * Represents an unspecified topic. This value should not be used.
     */
    ManagedTopicEnum["MANAGED_TOPIC_ENUM_UNSPECIFIED"] = "MANAGED_TOPIC_ENUM_UNSPECIFIED";
    /**
     * Represents significant personal information about the User like first names, relationships, hobbies, important dates.
     */
    ManagedTopicEnum["USER_PERSONAL_INFO"] = "USER_PERSONAL_INFO";
    /**
     * Represents stated or implied likes, dislikes, preferred styles, or patterns.
     */
    ManagedTopicEnum["USER_PREFERENCES"] = "USER_PREFERENCES";
    /**
     * Represents important milestones or conclusions within the dialogue.
     */
    ManagedTopicEnum["KEY_CONVERSATION_DETAILS"] = "KEY_CONVERSATION_DETAILS";
    /**
     * Represents information that the user explicitly requested to remember or forget.
     */
    ManagedTopicEnum["EXPLICIT_INSTRUCTIONS"] = "EXPLICIT_INSTRUCTIONS";
})(ManagedTopicEnum || (exports.ManagedTopicEnum = ManagedTopicEnum = {}));
/** The type of the memory. */
var MemoryType;
(function (MemoryType) {
    /**
     * Represents an unspecified memory type. This value should not be used.
     */
    MemoryType["MEMORY_TYPE_UNSPECIFIED"] = "MEMORY_TYPE_UNSPECIFIED";
    /**
     * Indicates belonging to a collection of natural language memories.
     */
    MemoryType["NATURAL_LANGUAGE_COLLECTION"] = "NATURAL_LANGUAGE_COLLECTION";
    /**
     * Indicates belonging to a structured profile.
     */
    MemoryType["STRUCTURED_PROFILE"] = "STRUCTURED_PROFILE";
})(MemoryType || (exports.MemoryType = MemoryType = {}));
/** Represents the operator to apply to the filter. If not set, then EQUAL will be used. */
var Operator;
(function (Operator) {
    /**
     * Represents an unspecified operator. Defaults to EQUAL.
     */
    Operator["OPERATOR_UNSPECIFIED"] = "OPERATOR_UNSPECIFIED";
    /**
     * Equal to.
     */
    Operator["EQUAL"] = "EQUAL";
    /**
     * Greater than.
     */
    Operator["GREATER_THAN"] = "GREATER_THAN";
    /**
     * Less than.
     */
    Operator["LESS_THAN"] = "LESS_THAN";
})(Operator || (exports.Operator = Operator = {}));
/** The coding language supported in this environment. */
var Language;
(function (Language) {
    /**
     * The default value. This value is unused.
     */
    Language["LANGUAGE_UNSPECIFIED"] = "LANGUAGE_UNSPECIFIED";
    /**
     * The coding language is Python.
     */
    Language["LANGUAGE_PYTHON"] = "LANGUAGE_PYTHON";
    /**
     * The coding language is JavaScript.
     */
    Language["LANGUAGE_JAVASCRIPT"] = "LANGUAGE_JAVASCRIPT";
})(Language || (exports.Language = Language = {}));
/** The machine config of the code execution environment. */
var MachineConfig;
(function (MachineConfig) {
    /**
     * The default value: milligcu 2000, memory 1.5Gib
     */
    MachineConfig["MACHINE_CONFIG_UNSPECIFIED"] = "MACHINE_CONFIG_UNSPECIFIED";
    /**
     * The default value: milligcu 4000, memory 4 Gib
     */
    MachineConfig["MACHINE_CONFIG_VCPU4_RAM4GIB"] = "MACHINE_CONFIG_VCPU4_RAM4GIB";
})(MachineConfig || (exports.MachineConfig = MachineConfig = {}));
/** Output only. The runtime state of the SandboxEnvironment. */
var State;
(function (State) {
    /**
     * The default value. This value is unused.
     */
    State["STATE_UNSPECIFIED"] = "STATE_UNSPECIFIED";
    /**
     * Runtime resources are being allocated for the sandbox environment.
     */
    State["STATE_PROVISIONING"] = "STATE_PROVISIONING";
    /**
     * Sandbox runtime is ready for serving.
     */
    State["STATE_RUNNING"] = "STATE_RUNNING";
    /**
     * Sandbox runtime is halted, performing tear down tasks.
     */
    State["STATE_DEPROVISIONING"] = "STATE_DEPROVISIONING";
    /**
     * Sandbox has terminated with underlying runtime failure.
     */
    State["STATE_TERMINATED"] = "STATE_TERMINATED";
    /**
     * Sandbox runtime has been deleted.
     */
    State["STATE_DELETED"] = "STATE_DELETED";
})(State || (exports.State = State = {}));
/** Framework used to build the application. */
var Framework;
(function (Framework) {
    /**
     * Unspecified framework.
     */
    Framework["FRAMEWORK_UNSPECIFIED"] = "FRAMEWORK_UNSPECIFIED";
    /**
     * React framework.
     */
    Framework["REACT"] = "REACT";
    /**
     * Angular framework.
     */
    Framework["ANGULAR"] = "ANGULAR";
})(Framework || (exports.Framework = Framework = {}));
/** The strategy to use when applying metadata to existing memories during consolidation. */
var MemoryMetadataMergeStrategy;
(function (MemoryMetadataMergeStrategy) {
    /**
     * The metadata merge strategy is unspecified.
     */
    MemoryMetadataMergeStrategy["METADATA_MERGE_STRATEGY_UNSPECIFIED"] = "METADATA_MERGE_STRATEGY_UNSPECIFIED";
    /**
     * Replace the metadata of the updated memories with the new metadata.
     */
    MemoryMetadataMergeStrategy["OVERWRITE"] = "OVERWRITE";
    /**
     * Append new metadata to the existing metadata. If there are duplicate keys, the existing values will be overwritten.
     */
    MemoryMetadataMergeStrategy["MERGE"] = "MERGE";
    /**
     * Restrict consolidation to memories that have exactly the same metadata as the request. If a memory doesn't have the same metadata, it is not eligible for consolidation.
     */
    MemoryMetadataMergeStrategy["REQUIRE_EXACT_MATCH"] = "REQUIRE_EXACT_MATCH";
})(MemoryMetadataMergeStrategy || (exports.MemoryMetadataMergeStrategy = MemoryMetadataMergeStrategy = {}));
/** The action to take. */
var GenerateMemoriesResponseGeneratedMemoryAction;
(function (GenerateMemoriesResponseGeneratedMemoryAction) {
    /**
     * The action is unspecified.
     */
    GenerateMemoriesResponseGeneratedMemoryAction["ACTION_UNSPECIFIED"] = "ACTION_UNSPECIFIED";
    /**
     * The memory was created.
     */
    GenerateMemoriesResponseGeneratedMemoryAction["CREATED"] = "CREATED";
    /**
     * The memory was updated. The `fact` field may not be updated if the existing fact is still accurate.
     */
    GenerateMemoriesResponseGeneratedMemoryAction["UPDATED"] = "UPDATED";
    /**
     * The memory was deleted.
     */
    GenerateMemoriesResponseGeneratedMemoryAction["DELETED"] = "DELETED";
})(GenerateMemoriesResponseGeneratedMemoryAction || (exports.GenerateMemoriesResponseGeneratedMemoryAction = GenerateMemoriesResponseGeneratedMemoryAction = {}));
/** Response for listing agent engines. */
class ListReasoningEnginesResponse {
}
exports.ListReasoningEnginesResponse = ListReasoningEnginesResponse;
/** The response for querying an agent engine. */
class QueryReasoningEngineResponse {
}
exports.QueryReasoningEngineResponse = QueryReasoningEngineResponse;
/** A memory that was generated. */
class GenerateMemoriesResponseGeneratedMemory {
}
exports.GenerateMemoriesResponseGeneratedMemory = GenerateMemoriesResponseGeneratedMemory;
/** The response for generating memories. */
class GenerateMemoriesResponse {
}
exports.GenerateMemoriesResponse = GenerateMemoriesResponse;
/** Response for listing agent engine memories. */
class ListReasoningEnginesMemoriesResponse {
}
exports.ListReasoningEnginesMemoriesResponse = ListReasoningEnginesMemoriesResponse;
/** A retrieved memory. */
class RetrieveMemoriesResponseRetrievedMemory {
}
exports.RetrieveMemoriesResponseRetrievedMemory = RetrieveMemoriesResponseRetrievedMemory;
/** The response for retrieving memories. */
class RetrieveMemoriesResponse {
}
exports.RetrieveMemoriesResponse = RetrieveMemoriesResponse;
/** The response for retrieving memory profiles. */
class RetrieveProfilesResponse {
}
exports.RetrieveProfilesResponse = RetrieveProfilesResponse;
/** The response for purging memories. */
class PurgeMemoriesResponse {
}
exports.PurgeMemoriesResponse = PurgeMemoriesResponse;
/** Response for listing agent engine memory revisions. */
class ListAgentEngineMemoryRevisionsResponse {
}
exports.ListAgentEngineMemoryRevisionsResponse = ListAgentEngineMemoryRevisionsResponse;
/** The response for executing a sandbox environment. */
class ExecuteSandboxEnvironmentResponse {
}
exports.ExecuteSandboxEnvironmentResponse = ExecuteSandboxEnvironmentResponse;
/** Response for listing agent engine sandboxes. */
class ListAgentEngineSandboxesResponse {
}
exports.ListAgentEngineSandboxesResponse = ListAgentEngineSandboxesResponse;
/** Response for listing agent engine sessions. */
class ListReasoningEnginesSessionsResponse {
}
exports.ListReasoningEnginesSessionsResponse = ListReasoningEnginesSessionsResponse;
/** Response for appending agent engine session event. */
class AppendAgentEngineSessionEventResponse {
}
exports.AppendAgentEngineSessionEventResponse = AppendAgentEngineSessionEventResponse;
/** Response for listing agent engine session events. */
class ListAgentEngineSessionEventsResponse {
}
exports.ListAgentEngineSessionEventsResponse = ListAgentEngineSessionEventsResponse;
/** Response from LRO. */
class CheckQueryJobResponse {
}
exports.CheckQueryJobResponse = CheckQueryJobResponse;
